crypto/fipsmodule/aes/aes_test.cc

```diff
@@ -125,6 +125,77 @@ static void TestKeyWrap(FileTest *t) {
                                ciphertext.data(), ciphertext.size()));
 }
 
+static void TestEVPKeyWrap(FileTest *t) {
+  std::vector<uint8_t> key, plaintext, ciphertext;
+  ASSERT_TRUE(t->GetBytes(&key, "Key"));
+  ASSERT_TRUE(t->GetBytes(&plaintext, "Plaintext"));
+  ASSERT_TRUE(t->GetBytes(&ciphertext, "Ciphertext"));
+
+  // Only 256 bit keys are supported for key wrap from EVP_CIPHER at the moment.
+  if (key.size() != 32) {
+    return;
+  }
+
+  const EVP_CIPHER *cipher = EVP_aes_256_wrap();
+
+  ASSERT_EQ(plaintext.size() + 8, ciphertext.size())
+      << "Invalid Plaintext and Ciphertext lengths.";
+
+  // Test encryption.
+  std::vector<uint8_t> out(ciphertext.size());
+  int len;
+  // Test with implicit IV.
+  bssl::ScopedEVP_CIPHER_CTX ctx;
+  ASSERT_TRUE(
+      EVP_EncryptInit_ex(ctx.get(), cipher, nullptr, key.data(), nullptr));
+  ASSERT_TRUE(EVP_EncryptUpdate(ctx.get(), out.data(), &len, plaintext.data(),
+                                plaintext.size()));
+  ASSERT_GE(len, 0);
+  ASSERT_TRUE(EVP_EncryptFinal(ctx.get(), out.data(), &len));
+  EXPECT_EQ(Bytes(ciphertext), Bytes(out));
+
+  // Test with explicit IV.
+  ctx.Reset();
+  ASSERT_TRUE(
+      EVP_EncryptInit_ex(ctx.get(), cipher, nullptr, key.data(), kDefaultIV));
+  ASSERT_TRUE(EVP_EncryptUpdate(ctx.get(), out.data(), &len, plaintext.data(),
+                                plaintext.size()));
+  ASSERT_GE(len, 0);
+  ASSERT_TRUE(EVP_EncryptFinal(ctx.get(), out.data(), &len));
+  EXPECT_EQ(Bytes(ciphertext), Bytes(out));
+
+  // Test decryption.
+  out.clear();
+  out.resize(plaintext.size());
+  ctx.Reset();
+  // Test with implicit IV.
+  ASSERT_TRUE(
+      EVP_DecryptInit_ex(ctx.get(), cipher, nullptr, key.data(), nullptr));
+  ASSERT_TRUE(EVP_DecryptUpdate(ctx.get(), out.data(), &len, ciphertext.data(),
+                                ciphertext.size()));
+  out.resize(len);
+  ASSERT_TRUE(EVP_EncryptFinal(ctx.get(), out.data(), &len));
+  EXPECT_EQ(Bytes(plaintext), Bytes(out));
+
+  // Test with explicit IV.
+  ctx.Reset();
+  ASSERT_TRUE(
+      EVP_DecryptInit_ex(ctx.get(), cipher, nullptr, key.data(), kDefaultIV));
+  ASSERT_TRUE(EVP_DecryptUpdate(ctx.get(), out.data(), &len, ciphertext.data(),
+                                ciphertext.size()));
+  out.resize(len);
+  ASSERT_TRUE(EVP_EncryptFinal(ctx.get(), out.data(), &len));
+  EXPECT_EQ(Bytes(plaintext), Bytes(out));
+
+  // Test corrupted ciphertext.
```

Comment:
why does `EVP_DecryptUpdate` fail with corrupted ciphertext?

Reply:
I think it's because the ciphertext can't be properly decrypted with same AES key here. In other words, this symmetric encryption scheme no longer recognizes the ciphertext with the same key. 
I took this from the test above in `TestKeyWrap`, which uses the AES Key `AES_wrap/unwrap_key` functions. Using `AES_unwrap_key` will also fail with a corrupted ciphertext.

Reply:
oh it's also authenticating... ok